docs(policies): add MeshTCPRoute docs (#1341)
kumahq/kuma-website: app/_src/policies/meshtcproute.md

{% tabs split useUrlFragment=false %}
{% tab split Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTCPRoute
metadata:
  name: tcp-route-1
  namespace: {{site.mesh_namespace}}
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend_kuma-demo_svc_8080
  to:
    - targetRef:
        kind: MeshService
        name: backend_kuma-demo_svc_3001
      rules:
        - default:
            backendRefs:
              - kind: MeshServiceSubset
                name: backend_kuma-demo_svc_3001
                tags:
                  version: "1.0"
                weight: 90
              - kind: MeshServiceSubset
                name: backend_kuma-demo_svc_3001
                tags:
                  version: "2.0"
                weight: 10
```

You can apply the configuration with `kubectl apply -f [..]`.
{% endtab %}

{% tab split Universal %}

```yaml
type: MeshTCPRoute
name: tcp-route-1
mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend_kuma-demo_svc_8080
  to:
    - targetRef:
        kind: MeshService
        name: backend_kuma-demo_svc_3001
      rules:

>         - matches:
>             - path:
>                 type: PathPrefix
>                 value: /api
>           default:

Reviewer: this part is missing from k8s equivalent so I'd remove that

```suggestion
        - default:
```

Reviewer 2: yeah it's definitely not part of MeshTCPRoute

---
title: MeshTCPRoute (beta)
---

{% warning %}
This policy uses a new policy matching algorithm and is in beta state. It shouldn't be combined with [TrafficRoute](../traffic-route).
{% endwarning %}

The `MeshTCPRoute` policy allows you to alter and redirect TCP requests
depending on where the request is coming from and where it's going to.

## TargetRef support matrix

| TargetRef type    | top level | to | from |
|-------------------|-----------|----|------|
| Mesh              | ✅         | ❌  | ❌    |
| MeshSubset        | ✅         | ❌  | ❌    |
| MeshService       | ✅         | ✅  | ❌    |
| MeshServiceSubset | ✅         | ❌  | ❌    |

> | MeshGatewayRoute  | ❌         | ❌  | ❌    |

Reviewer 2: ```suggestion
```

---
title: MeshTCPRoute (beta)
---

{% warning %}

> This policy uses new policy matching algorithm and is in beta state,
> it should not be mixed with [TrafficRoute](../traffic-route).

Reviewer 3: ```suggestion
This policy uses a new policy matching algorithm and is in beta state. It shouldn't be combined with [TrafficRoute](../traffic-route).
```

## TargetRef support matrix

| TargetRef type    | top level | to | from |
|-------------------|-----------|----|------|
| Mesh              | ✅         | ❌  | ❌    |
| MeshSubset        | ✅         | ❌  | ❌    |
| MeshService       | ✅         | ✅  | ❌    |
| MeshServiceSubset | ✅         | ❌  | ❌    |
| MeshGatewayRoute  | ❌         | ❌  | ❌    |

If you don't understand this table you should read 
[matching docs](/docs/{{ page.version }}/policies/targetref).

## Configuration

Unlike others outbound policies `MeshTCPRoute` doesn't contain `default`
directly in the `to` array. The `default` section is nested inside`rules`,
so the policy structure looks like this:

```yaml
spec:
  targetRef: # top-level targetRef selects a group of proxies to configure
    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset 
  to:
    - targetRef: # targetRef selects a destination (outbound listener)
        kind: MeshService
        name: backend
      rules:
        - default: # configuration applied for the matched TCP traffic
            backendRefs: [...]
```

### Default conf

- **`backendRefs`** - (optional) - list of destination for request to be
  redirected to

>   - **`kind`** - one of `MeshService`, `MeshServiceSubset`

Reviewer 3: ```suggestion
  - **`kind`**: Either `MeshService` or `MeshServiceSubset`
```

>   - **`name`** - service name

Reviewer 3: ```suggestion
  - **`name`**: The service name
```

>   - **`tags`** - service tags, must be specified if the `kind` is 
>     `MeshServiceSubset`

Reviewer 3: ```suggestion
  - **`tags`**: Service tags. These must be specified if the `kind` is 
    `MeshServiceSubset`.
```

  - **`weight`** - when a request matches the route, the choice of an upstream
    cluster is determined by its weight. Total weight is a sum of all weights
    in `backendRefs` list.

## Examples

### Traffic split

> We can use `MeshTCPRoute` to split a TCP traffic between services with
> different tags implementing A/B testing or canary deployments.

Reviewer 3: ```suggestion
You can use `MeshTCPRoute` to split TCP traffic between services with
different tags and implement A/B testing or canary deployments.
```

Reviewer 2: We already have `vale` set up, it may have broken at some point. I'll investigate it

Reviewer 3: @slonka @michaelbeaumont I'm not sure if Vale is checking for "we", but that's a really good suggestion. To get that added, the docs team might have to open a PR to add that rule to the Vale open source repo.

> Here is an example of a `MeshTCPRoute` that splits the traffic from 
> `frontend_kuma-demo_svc_8080` to `backend_kuma-demo_svc_3001` between versions.

Reviewer 3: ```suggestion
Here's an example of a `MeshTCPRoute` that splits the traffic from 
`frontend_kuma-demo_svc_8080` to `backend_kuma-demo_svc_3001` between versions:
```

{% tabs split useUrlFragment=false %}
{% tab split Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTCPRoute
metadata:
  name: tcp-route-1
  namespace: {{site.mesh_namespace}}
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend_kuma-demo_svc_8080
  to:
    - targetRef:
        kind: MeshService
        name: backend_kuma-demo_svc_3001
      rules:
        - default:
            backendRefs:
              - kind: MeshServiceSubset
                name: backend_kuma-demo_svc_3001
                tags:
                  version: "1.0"
                weight: 90
              - kind: MeshServiceSubset
                name: backend_kuma-demo_svc_3001
                tags:
                  version: "2.0"
                weight: 10
```

> We will apply the configuration with `kubectl apply -f [..]`.

Reviewer 3: ```suggestion
You can apply the configuration with `kubectl apply -f [..]`.
```

{% endtab %}

{% tab split Universal %}

```yaml
type: MeshTCPRoute
name: tcp-route-1
mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend_kuma-demo_svc_8080
  to:
    - targetRef:
        kind: MeshService
        name: backend_kuma-demo_svc_3001
      rules:
        - matches:
            - path:
                type: PathPrefix
                value: /api
          default:
            backendRefs:
              - kind: MeshServiceSubset
                name: backend_kuma-demo_svc_3001
                tags:
                  version: "1.0"
                weight: 90
              - kind: MeshServiceSubset
                name: backend_kuma-demo_svc_3001
                tags:
                  version: "2.0"
                weight: 10
```

We will apply the configuration with `kumactl apply -f [..]` or via
the [HTTP API](/docs/{{ page.version }}/reference/http-api).
{% endtab %}
{% endtabs %}

### Traffic redirection

We can use `MeshTCPRoute` to redirect outgoing traffic from one service to
another.

> Here is an example of a `MeshTCPRoute` that redirects outgoing traffic 
> originated at `frontend_kuma-demo_svc_8080` from `backend_kuma-demo_svc_3001`
> to `external-backend`. 

Reviewer 3: ```suggestion
Here's an example of a `MeshTCPRoute` that redirects outgoing traffic 
originating at `frontend_kuma-demo_svc_8080` from `backend_kuma-demo_svc_3001`
to `external-backend`:
```

{% tabs modifications useUrlFragment=false %}
{% tab modifications Kubernetes %}

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTCPRoute
metadata:
  name: tcp-route-1
  namespace: {{site.mesh_namespace}}
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend_kuma-demo_svc_8080
  to:
    - targetRef:
        kind: MeshService
        name: backend_kuma-demo_svc_3001
      rules:
        - default:
            backendRefs:
              - kind: MeshService
                name: external-backend
```

> We will apply the configuration with `kubectl apply -f [..]`.

Reviewer 3: ```suggestion
You can apply the configuration with `kubectl apply -f [..]`.
```

{% endtab %}

{% tab modifications Universal %}

```yaml
type: MeshTCPRoute
name: tcp-route-1
mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend_kuma-demo_svc_8080
  to:
    - targetRef:
        kind: MeshService
        name: backend_kuma-demo_svc_3001
      rules:
        - default:
            backendRefs:
              - kind: MeshService
                name: external-backend
```

> We will apply the configuration with `kumactl apply -f [..]` or via
> the [HTTP API](/docs/{{ page.version }}/reference/http-api).

Reviewer 3: ```suggestion
You can apply the configuration with `kumactl apply -f [..]` or use
the [HTTP API](/docs/{{ page.version }}/reference/http-api).
```

{% endtab %}
{% endtabs %}

## Route policies with different types targeting the same destination

> If multiple route policies with different types (`MeshTCPRoute`, `MeshHTTPRoute`
> etc.) target the same destination, only a single route type with the highest
> specificity will be applied.

Reviewer 3: ```suggestion
If multiple route policies with different types (`MeshTCPRoute` and `MeshHTTPRoute`
for example) target the same destination, only a single route type with the highest
specificity will be applied.
```

In example, in a situation when both `MeshTCPRoute` and `MeshHTTPRoute` exist,
and target the same destination, like below:

**MeshTCPRoute**:
```yaml
# [...]
targetRef:
  kind: MeshService
  name: frontend
to:
  - targetRef:
      kind: MeshService
      name: backend
    rules:
      - default:
          backendRefs:
            - kind: MeshService
              name: other-tcp-backend
```

> **MeshHTTPRoute**

Reviewer 3: ```suggestion
**MeshHTTPRoute**:
```

```yaml
# [...]
targetRef:
  kind: MeshService
  name: frontend
to:
  - targetRef:
      kind: MeshService
      name: backend
    rules:
      - matches:
          - path:
              type: PathPrefix
              value: "/"
        default:
          backendRefs:
            - kind: MeshService
              name: other-http-backend
```

> Depending on the `backend`'s protocol:
> - `MeshHTTPRoute` will be applied in case of `http`, `http2` or `grpc`
> - `MeshTCPRoute` will be applied in case of `tcp`, `kafka` or when unspecified 

Reviewer 3: ```suggestion
Depending on the `backend`'s protocol:
- `MeshHTTPRoute` will be applied if `http`, `http2`, or `grpc` are specified
- `MeshTCPRoute` will be applied if `tcp` or `kafka` is specified, or when nothing is specified 
```
